Adding a spline joint to the joint collection

MegMll opened the pull request:
## Description

In this PR, we introduce the new spline joint to pinocchio. This implementation is based on the paper of Lee et al. available [here](https://web.cs.ucla.edu/~dt/papers/siggraph08/siggraph08.pdf)

It allows to define a number of SE3 frames, that will define the movement and dynamics of the joint. It's very useful for biomechanics, but can be used in other applications. 

Here are videos of two examples written by @lpuch
- Knee
[Knee Movement](https://github.com/user-attachments/assets/67384bce-1384-4123-8523-8907f4be8c24)


- Random Spline
[Random Spline](https://github.com/user-attachments/assets/587de626-cb73-4fd6-8e6e-04b1cec54a9d)


## Future 
- [ ] Introduce the [Spline Surface](https://web.mit.edu/hyperbook/Patrikalakis-Maekawa-Cho/node19.html)

MegMll committed:
Joints - Add a first implementation of joint spline

MegMll committed:
Joints - Add a first implementation of joint spline

MegMll committed:
Unittest - Add TU for spline joint

MegMll committed:
Unittest - Add TU for spline joint

MegMll committed:
Bindings - add bindings for spline joint

MegMll committed:
Bindings - add bindings for spline joint

MegMll committed:
Spatial - Weird fix to make spline joint work for now (to remove)

MegMll committed:
Spatial - Weird fix to make spline joint work for now (to remove)

MegMll committed:
Unittest - update joint unittests to include spline joints

MegMll committed:
Unittest - update joint unittests to include spline joints

MegMll committed:
SplineJoint - apply pre-commit

MegMll committed:
SplineJoint - apply pre-commit

MegMll committed:
Spline - use the true cumulative form of the spline

MegMll committed:
Spline - use the true cumulative form of the spline

MegMll committed:
Spline - Add isEqual Method

MegMll committed:
Spline - Add isEqual Method

MegMll committed:
Spline - Add TU with finite difference

MegMll committed:
Spline - Add TU with finite difference

MegMll committed:
Unittest - fix tests for spline joint

MegMll committed:
Unittest - fix tests for spline joint

MegMll committed:
Spline joint - change if/else to be casadi compatible

MegMll committed:
Spline joint - change if/else to be casadi compatible

MegMll committed:
Joint Spline - Remove useless comments

MegMll committed:
Joint Spline - Remove useless comments

MegMll committed:
Joint Spline - Fix indexing

MegMll committed:
Joint Spline - Fix indexing

MegMll committed:
Joint Spline - Eigen No Malloc checked and passed

MegMll committed:
Joint Spline - Eigen No Malloc checked and passed

MegMll committed:
Spline joints - fix casadi joints tests

MegMll committed:
Spline joints - fix casadi joints tests

MegMll committed:
Joint Spline - Add a build method

MegMll committed:
Joint Spline - Add a build method

MegMll committed:
Unittest - Fix joint spline's unittests

MegMll committed:
Unittest - Fix joint spline's unittests

jorisv commented on include/pinocchio/src/multibody/joint/joint-collection.hxx:
```diff
@@ -106,6 +109,7 @@ namespace pinocchio
       JointModelHz,
       JointModelHelicalUnaligned,
       JointModelUniversal,
+      JointModelSpline,
```
Do we need it in the default joint collection ?

MegMll replied:
I'm almost certain it will be used with casadi and double for biomechanics application. But maybe we can put it somewhere else, if you think it's better

jorisv commented on include/pinocchio/multibody/joint/joint-spline.hpp:
```diff
@@ -0,0 +1,557 @@
+#include "pinocchio/math/fwd.hpp"
+#include "pinocchio/math/quaternion.hpp"
+
+#include <iostream>
```
```suggestion
#include "pinocchio/eigen-macros.hpp"
#include "pinocchio/spatial/explog.hpp"
#include "pinocchio/multibody/joint/joint-base.hpp"
#include "pinocchio/multibody/joint-motion-subspace.hpp"

```

jorisv commented on include/pinocchio/multibody/joint/joint-spline.hpp:
```diff
@@ -0,0 +1,557 @@
+    EIGEN_MAKE_ALIGNED_OPERATOR_NEW
+    typedef JointSplineTpl<_Scalar, _Options> JointDerived;
+    typedef SE3Tpl<_Scalar, _Options> SE3;
+    typedef MotionTpl<_Scalar, _Options> Motion;
```
`Transformation_t` and `Motion_t` are already defined in `PINOCCHIO_JOINT_TYPEDEF_TEMPLATE`

MegMll committed:
Model-graph - Add spline joint to model graph

MegMll committed:
Model-graph - Add spline joint to model graph

jorisv commented on include/pinocchio/multibody/joint/joint-spline.hpp:
```diff
@@ -0,0 +1,557 @@
+      makeKnots();
+      computeRelativeMotions();
+    }
+
```
This is a little error prone. `JointModelSplineTpl` instance will not work until you call this function.

Two way of fixing that:
1. Remove `addControlFrame`, add a `setControlFrames` function that will set the full vector and call buildJoint (now private). Do the same on the constructor.
2. Add a `JointModelSplineBuilderTpl` class that you can edit, then `JointModelSplineBuilderTpl::buildJoints` will return a well configured `JointModelSplineTpl`.

Option 1. is way simpler. I think, user can manage by themself to keep a ctrlFrames vector on their side.

jorisv commented on include/pinocchio/multibody/joint/joint-spline.hpp:
```diff
@@ -0,0 +1,557 @@
+    }
+
+    template<typename ConfigVector>
+    void calc(JointDataDerived & data, const typename Eigen::MatrixBase<ConfigVector> & qs) const
```
```suggestion
    void calc(JointDataDerived & data, const Eigen::MatrixBase<ConfigVector> & qs) const
```

jorisv commented on include/pinocchio/multibody/joint/joint-spline.hpp:
```diff
@@ -0,0 +1,557 @@
+    void calc(
+      JointDataDerived & data,
+      const typename Eigen::MatrixBase<ConfigVector> & qs,
+      const typename Eigen::MatrixBase<TangentVector> & vs) const
```
```suggestion
      const Eigen::MatrixBase<ConfigVector> & qs,
      const Eigen::MatrixBase<TangentVector> & vs) const
```

jorisv commented on include/pinocchio/multibody/joint/joint-spline.hpp:
```diff
@@ -0,0 +1,557 @@
+    void calc(
+      JointDataDerived & data,
+      const Blank not_used,
+      const typename Eigen::MatrixBase<TangentVector> & vs) const
```
```suggestion
      const Eigen::MatrixBase<TangentVector> & vs) const
```

jorisv commented on include/pinocchio/src/multibody/joint/joint-spline.hxx:
```diff
@@ -0,0 +1,557 @@
+      // Basis functions and their derivatives
+      data.N.setZero();
+      data.N_der.setZero();
+      for (size_t i = indexes.start_idx; i < indexes.end_idx; i++)
```
`start_idx` is an int where `size_t` is unsigned.
Since `start_idx` and `end_idx` can't be negative, I think it's better to turn them into size_t.

jorisv commented on include/pinocchio/src/multibody/joint/joint-spline.hxx:
```diff
@@ -0,0 +1,557 @@
+      const int n_knots = nbCtrlFrames + degree + 1;
+      knots.resize(n_knots);
+      knots.head(degree + 1).setZero();
+      const Scalar denominator = static_cast<Scalar>(nbCtrlFrames - degree + 1);
```
Whe should trigger an exception if the user don't provide enough control point

jorisv replied:
Put some guards on degree size vs control points numbers

jorisv commented on include/pinocchio/multibody/joint/joint-spline.hpp:
```diff
@@ -0,0 +1,557 @@
+
+        data.M = data.M * exp6(relativeMotions[i - 1] * phi_i);
+        data.S.matrix() = exp6(relativeMotions[i - 1] * phi_i).actInv(data.S)
+                          + relativeMotions[i - 1].toVector() * phi_dot_i;
```
Compiler will probably optimize it, but you can store `exp6(relativeMotions[i - 1] * phi_i)` in a temporary variable.

jorisv commented on include/pinocchio/multibody/joint/joint-spline.hpp:
```diff
@@ -0,0 +1,557 @@
+          + exp6(relativeMotions[i - 1] * phi_i)
+              .actInv(data.c + Motion(data.S.matrix()).cross(relativeMotions[i - 1]) * phi_dot_i);
+        data.S.matrix() = exp6(relativeMotions[i - 1] * phi_i).actInv(data.S)
+                          + relativeMotions[i - 1].toVector() * phi_dot_i;
```
Same here, `exp6(relativeMotions[i - 1] * phi_i)` can be stored somewhere. This will help readability.

jorisv commented on include/pinocchio/multibody/joint/joint-spline.hpp:
```diff
@@ -0,0 +1,557 @@
+          + exp6(relativeMotions[i] * phi_i)
+              .actInv(data.c + Motion(data.S.matrix()).cross(relativeMotions[i]) * phi_dot_i);
+        data.S.matrix() = exp6(relativeMotions[i] * phi_i).actInv(data.S)
+                          + relativeMotions[i].toVector() * phi_dot_i;
```
Store `exp6(relativeMotions[i] * phi_i)` in a temporary.

jorisv commented on include/pinocchio/multibody/joint/joint-spline.hpp:
```diff
@@ -0,0 +1,557 @@
+          pinocchio::internal::LE, knots[i], x,
+          pinocchio::internal::if_then_else(
+            pinocchio::internal::LE, x, knots[i + 1], Scalar(1), Scalar(0)),
+          Scalar(0));
```
Oh my god !

1.: You can do a `using internal::if_then_else;` at the beggining of the function ?
2.: Can you write the condition written normally in comment ? Like [here](https://github.com/stack-of-tasks/pinocchio/blob/devel/include/pinocchio/multibody/liegroup/special-orthogonal.hpp#L87-L93). You can use `// clang-format off` and `// clang-format on` comment around the commented code to keep a nice formatting.

MegMll replied:
Yeah sorry, it's unreadable....

jorisv commented on include/pinocchio/multibody/joint/joint-spline.hpp:
```diff
@@ -0,0 +1,557 @@
+  };
+
+  template<int Options>
+  struct FindSpan<double, Options>
```
double specialization will not work on float or other true scalar type.
Also, maybe some custom scalar can still extract true scalar and use the normal function.

So it's better to provide a default implementation and specialize for custom scalar in dedicated header as done [here](https://github.com/stack-of-tasks/pinocchio/blob/devel/include/pinocchio/autodiff/cppad/algorithm/aba.hpp)

MegMll replied:
only casadi seemed to cause problems here, so i only specialized it for casadi

jorisv commented on include/pinocchio/src/serialization/joints-data.hxx:
```diff
@@ -269,6 +269,16 @@ namespace boost
+      pinocchio::JointDataSplineTpl<Scalar, Options> & joint,
+      const unsigned int version)
+    {
+      typedef pinocchio::JointDataSplineTpl<Scalar, Options> JointType;
```
You forget the bspline values:

```cpp
    Vector N;
    Vector N_der;
    Vector N_der2;
```

jorisv commented on include/pinocchio/src/serialization/joints-model.hxx:
```diff
@@ -311,6 +311,19 @@ namespace boost
+      typedef pinocchio::JointModelSplineTpl<Scalar, Options> JointType;
+      ar & make_nvp("ctrlFrames", joint.ctrlFrames);
+      ar & make_nvp("degree", joint.degree);
+      ar & make_nvp("nbCtrlFrames", joint.nbCtrlFrames);
```
Did this will call `buildJoint` ?

MegMll replied:
no indeed, so i should probably serialize knots vector and relativeMotions

MegMll committed:
JointSpline - make some correction to the joint

MegMll committed:
JointSpline - make some correction to the joint

MegMll committed:
TU - apply new jointSpline constructor

MegMll committed:
TU - apply new jointSpline constructor

MegMll committed:
ModelGraph - use new JointSpline method

MegMll committed:
ModelGraph - use new JointSpline method

MegMll committed:
Serialization - Add attributes from spline model and data

MegMll committed:
Serialization - Add attributes from spline model and data

MegMll committed:
Unittest - use std vector instead of se3

MegMll committed:
Unittest - use std vector instead of se3

MegMll committed:
JointSpline - Diverse Changes

MegMll committed:
JointSpline - Diverse Changes

MegMll committed:
Casadi - Add a separated file for findSpan to be able to do specialization for casadi

MegMll committed:
Casadi - Add a separated file for findSpan to be able to do specialization for casadi

MegMll committed:
Bindings - modify joint spline binding

MegMll committed:
Bindings - modify joint spline binding

MegMll committed:
JointSpline - Remove findSpan

MegMll committed:
JointSpline - Remove findSpan

MegMll committed:
Unittest - remove buildJoint function

MegMll committed:
Unittest - remove buildJoint function

MegMll committed:
ModelGraph - finish adding jointSpline to model graph + unittest

MegMll committed:
ModelGraph - finish adding jointSpline to model graph + unittest

jorisv committed:
cppadcg: Fix cppadcg cast test with spline joint

Spline joint will call `log6` method in the cast method.
This method is unsupported with Cg<double> scalar type.
We use Ad<Cg<double>> type instead, that is well supported.

jorisv committed:
cppadcg: Fix cppadcg cast test with spline joint

Spline joint will call `log6` method in the cast method.
This method is unsupported with Cg<double> scalar type.
We use Ad<Cg<double>> type instead, that is well supported.

jorisv committed:
multiprecision: Fix multiprecision cast test with Spline joint

`log6` call is not supported with `boost::multiprecision`.
We force an expression evaluation into a scalar to fix that.

jorisv committed:
multiprecision: Fix multiprecision cast test with Spline joint

`log6` call is not supported with `boost::multiprecision`.
We force an expression evaluation into a scalar to fix that.

jorisv committed:
multiprecision: Add missing header

jorisv committed:
multiprecision: Add missing header

jorisv committed:
lint: Run linter

jorisv committed:
lint: Run linter

MegMll committed:
JointSpline - Removed comments

MegMll committed:
JointSpline - Removed comments

MegMll committed:
JointSpline - move bspline functions in public to be able to test it

MegMll committed:
JointSpline - move bspline functions in public to be able to test it

MegMll committed:
Unittest - add more test for joint spline

MegMll committed:
Unittest - add more test for joint spline

MegMll committed:
Bindings - Add another constructor for joint spline

MegMll committed:
Bindings - Add another constructor for joint spline

MegMll committed:
Unittest - put full knee trajectory

MegMll committed:
Unittest - put full knee trajectory

MegMll committed:
Unittest - fix getTrajectory function

MegMll committed:
Unittest - fix getTrajectory function

MegMll committed:
Bindings - add modelgraph bindings constant

MegMll committed:
Bindings - add modelgraph bindings constant

MegMll committed:
JointSpline - Fix indexing

MegMll committed:
JointSpline - Fix indexing

MegMll committed:
JointSpline - Add a doc

MegMll committed:
JointSpline - Add a doc

MegMll committed:
JointSpline - Fix bias computation

MegMll committed:
JointSpline - Fix bias computation

MegMll committed:
Unittest - add a finite difference test for S and bias c for JointSpline

MegMll committed:
Unittest - add a finite difference test for S and bias c for JointSpline

MegMll committed:
JointSpline - Add comments and complete isEqual function

MegMll committed:
JointSpline - Add comments and complete isEqual function

MegMll committed:
Unittest - Try to add tolerance to see if it fixes macos-intel problem

MegMll committed:
Unittest - Try to add tolerance to see if it fixes macos-intel problem

MegMll committed:
Unittest - fix compilation error

MegMll committed:
Unittest - fix compilation error

MegMll committed:
JointSpline - add assert for configuration to be between 0 and 1

MegMll committed:
JointSpline - add assert for configuration to be between 0 and 1

MegMll committed:
Bindings - Add graph spline joints bindings

MegMll committed:
Bindings - Add graph spline joints bindings

MegMll committed:
Algo/Splines - remove comments

MegMll committed:
Algo/Splines - remove comments

MegMll committed:
Unittest - add find span test for q = 0

MegMll committed:
Unittest - add find span test for q = 0

MegMll committed:
Changelog - update

MegMll committed:
Changelog - update

MegMll committed:
JointSpline - remove SE3 and use Transformation_t

MegMll committed:
JointSpline - remove SE3 and use Transformation_t

MegMll committed:
Unittest - remove comments

MegMll committed:
Unittest - remove comments

MegMll committed:
JointSpline - Use transformation_t and motion_t

MegMll committed:
JointSpline - Use transformation_t and motion_t

MegMll committed:
Examples - Add a small examples on what the spline joint can do

MegMll committed:
Examples - Add a small examples on what the spline joint can do

MegMll committed:
Unittest - Specialize for joint spline

MegMll committed:
Unittest - Specialize for joint spline

MegMll committed:
Examples - add spline joint examples

MegMll committed:
Examples - add spline joint examples

MegMll committed:
Unittest - Fix joint generic test for joint spline

MegMll committed:
Unittest - Fix joint generic test for joint spline

MegMll committed:
Examples - Fix some linting errors

MegMll committed:
Examples - Fix some linting errors

MegMll committed:
JointSpline - Disable assert

MegMll committed:
JointSpline - Disable assert

jorisv commented on include/pinocchio/algorithm/splines.hpp:
```diff
@@ -0,0 +1,49 @@
+//
+
+#ifndef __pinocchio_algorithm_splines_hpp__
+#define __pinocchio_algorithm_splines_hpp__
+
+namespace pinocchio
+{
+  struct SpanIndexes
+  {
+    size_t start_idx;
+    size_t end_idx;
+  };
+
+  template<typename Scalar, int Options>
+  struct FindSpan
```
Add a doc string to explain what are doing these structures.

jorisv commented on include/pinocchio/src/parsers/graph/graph-visitor.hxx:
```diff
@@ -129,6 +129,13 @@ namespace pinocchio
             "Graph - Joint Mimic cannot have a q_ref. Please use the joint offset directly.");
         }
+
+        SE3 operator()(const JointSpline &) const
+        {
+          PINOCCHIO_THROW_PRETTY(
+            std::invalid_argument,
+            "Graph - Joint Spline cannot have a q_ref. Please use the joint offset directly.");
+        }
```
Why ? It's strange but I don't get why it's not possible.

MegMll replied:
I looked into it, and for now the spline joint can not accept q_ref. if we put a q_ref, that means we need to redefine the spline joint, from this q_ref to the end, or make the q_ref the 0 of the joint, and compute position from both side of the q_ref (-0.5/0.5, for example).
So q_ref is for now, not possible, but could be integrated when the fit_spline algorithm is added or when the upper and lower bounds are implemented

jorisv commented on src/parsers/graph/model-graph-algo.cpp:
```diff
@@ -338,6 +348,28 @@ namespace pinocchio
+        {
+          if (!edge.forward)
+            PINOCCHIO_THROW_PRETTY(
+              std::invalid_argument, "Graph - JointSpline cannot be reversed.");
```
We have to think about that.

Right now I'm not very confident about the reversal of this joint.
I don't think creating a ctrlFrames vector with inverse transformation inside will give us the inverse transformation.

MegMll replied:
I looked into it as well. Indeed, reversing the ctrlFrames vector will only work when degree of the spline is 1. 
So maybe it's better to leave it as it is, right now.

jorisv commented on unittest/casadi/joints.cpp:
```diff
@@ -285,6 +285,25 @@ struct init<pinocchio::JointModelMimicTpl<Scalar, Options, JointCollection>>
+    PINOCCHIO_ALIGNED_STD_VECTOR(SE3) ctrlFrames;
+    ctrlFrames.push_back(SE3::Identity());
+    ctrlFrames.push_back(SE3(Matrix3s::Identity(), Vector3s(Scalar(0.), Scalar(0.), Scalar(1.))));
+    JointModel jmodel(ctrlFrames, 1);
```
We should create a degree 3 spline so `bsplineBasisDerivative` and `bsplineBasisDerivative2` are called.

jorisv commented on unittest/casadi/joints.cpp:
```diff
@@ -359,6 +378,24 @@ struct TestADOnJoints
+    PINOCCHIO_ALIGNED_STD_VECTOR(SE3) ctrlFrames;
+    ctrlFrames.push_back(SE3::Identity());
+    ctrlFrames.push_back(SE3(Matrix3s::Identity(), Vector3s(Scalar(0.), Scalar(0.), Scalar(1.))));
+    JointModel jmodel(ctrlFrames, 1);
```
We should create a degree 3 spline so `bsplineBasisDerivative` and `bsplineBasisDerivative2` are called.

jorisv commented on unittest/cppad/joints.cpp:
```diff
@@ -189,6 +189,23 @@ struct TestADOnJoints
+    PINOCCHIO_ALIGNED_STD_VECTOR(SE3) ctrlFrames;
+    ctrlFrames.push_back(SE3::Identity());
+    ctrlFrames.push_back(SE3(Matrix3s::Identity(), Vector3s(Scalar(0.), Scalar(0.), Scalar(1.))));
+    JointModel jmodel(ctrlFrames, 1);
```
We should create a degree 3 spline so `bsplineBasisDerivative` and `bsplineBasisDerivative2` are called.

jorisv commented on unittest/finite-differences.cpp:
```diff
@@ -212,6 +212,22 @@ struct init<pinocchio::JointModelMimicTpl<Scalar, Options, JointCollection>>
+    PINOCCHIO_ALIGNED_STD_VECTOR(pinocchio::SE3) ctrlFrames;
+    ctrlFrames.push_back(pinocchio::SE3::Identity());
+    ctrlFrames.push_back(pinocchio::SE3(Eigen::Matrix3d::Identity(), Eigen::Vector3d(0., 0., 1.)));
+    JointModel jmodel(ctrlFrames, 1);
```
We should create a degree 3 spline so `bsplineBasisDerivative` and `bsplineBasisDerivative2` are called.

jorisv committed:
core: Apply small changes

jorisv committed:
core: Apply small changes

jorisv commented on unittest/casadi/joints.cpp:
```diff
@@ -390,6 +427,145 @@ struct TestADOnJoints
+    ConfigVector q(jmodel.nq());
+
+    ConfigVector lb(ConfigVector::Constant(jmodel.nq(), 0.));
+    ConfigVector ub(ConfigVector::Constant(jmodel.nq(), 1.));
```
The only difference between generic `test` function and specialized one for `JointModelSplineTpl` is the configuration bound.
Is there way to specialize RandomConfigurationStep to handle that ?
If there is no way, can generate the bounds from `TestADOnJoints` visitor to avoid duplicating the code.

MegMll replied:
we can specialize the RandomConfigurationStep, like it's done for CompositeJoint, indeed. I'll do it, and you let me know, what you think.

jorisv commented on unittest/finite-differences.cpp:
```diff
@@ -222,6 +238,46 @@ struct FiniteDiffJoint
+    std::cout << "name: " << jmodel.classname() << std::endl;
+    JointDataSpline jdata = jmodel.createData();
+
+    CV q = LieGroupType().randomConfiguration(CV::Zero(), CV::Ones());
```
The only difference here is randomConfiguration generation. Is there a way to handle it in LieGroup ?
Otherwise, it could be nice to generate it from the visitor to avoid duplicating the code.

MegMll replied:
i checked, and it requires a new lieGroup, from my understanding. So i will generate bounds from the visitors

jorisv commented on unittest/joint-generic.cpp:
```diff
@@ -75,7 +75,122 @@ void test_joint_methods(
+  Eigen::VectorXd armature =
+    Eigen::VectorXd::Random(jdata.S().nv()) + Eigen::VectorXd::Ones(jdata.S().nv());
+
+  q1 = LieGroupType().randomConfiguration(
+    Eigen::VectorXd::Zero(jmodel.nq()), Eigen::VectorXd::Ones(jmodel.nq()));
+  q2 = LieGroupType().randomConfiguration(
+    Eigen::VectorXd::Zero(jmodel.nq()), Eigen::VectorXd::Ones(jmodel.nq()));
```
Same here, randomConfiguration is the only difference between generic and spline joint.
If there is no way to specialize randomConfiguration, we should add a mechanism to this test to generate a valid randomConfiguration without duplicating the test.

jorisv commented on unittest/joint-generic.cpp:
```diff
@@ -279,6 +394,22 @@ struct init<pinocchio::JointModelHelicalUnalignedTpl<Scalar, Options>>
+    PINOCCHIO_ALIGNED_STD_VECTOR(SE3) ctrlFrames;
+    ctrlFrames.push_back(SE3::Identity());
+    ctrlFrames.push_back(SE3(Eigen::Matrix3d::Identity(), Eigen::Vector3d(0., 0., 1.)));
+    JointModel jmodel(ctrlFrames, 1);
```
We should create an order 3 splines to test all the joint.

jorisv commented on unittest/joint-spline.cpp:
```diff
@@ -0,0 +1,422 @@
+
+  // Check Values
+  Eigen::VectorXd knots_expected(degree + ctrlFrames.size() + 1);
+  knots_expected << 0., 0., 0., 0., 1., 1., 1., 1.;
```
Add more control frame to test intermediate knots construction.

jorisv commented on unittest/joint-spline.cpp:
```diff
@@ -0,0 +1,422 @@
+  // check values
+  for (size_t i = 0; i < ctrlFrames.size() - 1; i++)
+    BOOST_CHECK(jmodel.relativeMotions[i].isApprox(
+      pinocchio::log6(ctrlFrames[i].inverse() * ctrlFrames[i + 1])));
```
This test duplicate the `computeRelativeMotions` code.

To avoid that and have a more robust test we can generate ctrlFrames from random `Motion` and check that `jmodel.relativeMotions` contains the generated random `Motion`

jorisv commented on include/pinocchio/src/multibody/joint/joint-spline.hxx:
Is there a way that `neutral` and `randomConfiguration` return us a valid spline configuration ?

MegMll replied:
neutral is working, because it returns 0. However if we want random, we need to redefine a lieGroup...

jorisv commented on include/pinocchio/src/multibody/joint/joint-spline.hxx:
Do we want to manage spline with abritrary bounds ?

MegMll replied:
in the future, it would be nice

MegMll committed:
Algo/splines - add doc

MegMll committed:
Algo/splines - add doc

MegMll committed:
Unittest - Avoid code duplication for randomConfig vector generation

MegMll committed:
Unittest - Avoid code duplication for randomConfig vector generation

MegMll committed:
Unittest - more complex and complete tests

MegMll committed:
Unittest - more complex and complete tests

jorisv commented on unittest/finite-differences.cpp:
```diff
@@ -61,154 +61,231 @@ void filterValue(MatrixBase<Matrix> & mat, typename Matrix::Scalar value)
 
 template<typename JointModel_>
-struct init;
+struct JmodelWithBounds
```
Maybe we should use a more generic name (if with want to add more attribute in the future).
Maybe JointModelWithParameters.

jorisv commented on unittest/finite-differences.cpp:
```diff
@@ -229,14 +306,18 @@ struct FiniteDiffJoint
     typedef typename LieGroup<JointModel>::type LieGroupType;
 
-    JointModel jmodel = init<JointModel>::run();
+    JmodelWithBounds jmodelWithBounds = init<JointModel>::run();
```
```suggestion
    auto jmodelWithBounds = init<JointModel>::run();
```
Some compiler didn't like the missing template argument in the return type.

jorisv commented on unittest/joint-generic.cpp:
This test can fail because on bad range bound with spline joint (seen on OSX)

MegMll committed:
Unittest/cppad - use degree 3 spline joint

MegMll committed:
Unittest/cppad - use degree 3 spline joint

MegMll committed:
unittest - change name of struct to be more generic

MegMll committed:
unittest - change name of struct to be more generic

MegMll committed:
unittest - avoid code duplication

MegMll committed:
unittest - avoid code duplication

MegMll committed:
unittest - avoid code duplication

MegMll committed:
unittest - avoid code duplication

MegMll committed:
Unittest - fix casadi/joints test

MegMll committed:
Unittest - fix casadi/joints test

MegMll committed:
Unittest - add forgotten helical unaligned joint

MegMll committed:
Unittest - add forgotten helical unaligned joint

MegMll committed:
Unittest - Fix casadi test for universal joints

MegMll committed:
Unittest - Fix casadi test for universal joints

MegMll committed:
Mode-graph - use new function for joint spline

MegMll committed:
Mode-graph - use new function for joint spline

florent-lamiraux commented:
Please be aware that modifying the joint collection breaks the API.

jorisv commented:
@florent-lamiraux We are not considering adding a joint an API break.
If your code rely on JointCollection order or must handle all the joints inside it, please, create your own joint collection.

MegMll committed:
Linting

MegMll committed:
Linting

MegMll committed:
unittest - fix new joint

MegMll committed:
unittest - fix new joint

MegMll committed:
JointSpline - Fix bias computation

MegMll committed:
JointSpline - Fix bias computation

MegMll committed:
Unittest - Fix finite-difference test

MegMll committed:
Unittest - Fix finite-difference test

MegMll committed:
unittest - use config vector of joints

MegMll committed:
unittest - use config vector of joints

MegMll committed:
unittest - remove useless lines

MegMll committed:
unittest - remove useless lines

MegMll committed:
Unittest - Add a utils file to centralize init of joints with parameters

MegMll committed:
Unittest - Add a utils file to centralize init of joints with parameters

MegMll committed:
Run pre-commit

MegMll committed:
Run pre-commit

MegMll committed:
unittest/utils - Fix mimic init

MegMll committed:
unittest/utils - Fix mimic init

Ipuch commented:
@MegMll and I were talking about this PR. As Megan left WILLOW and she is full time on other projects now, it's hard for her to pursue the debugging that this PR requires, and looking for why the tests were not passing. Does anyone would like to help and/or review this PR?

Here is the focused help, that would be appreciated and could unlock our next developments.

- We don't know why window tests with python doesn't pass ?
- We don't know why it compiles with MacOs Arm, but the test doesn't pass when executed ?

Let us know in which time frame anyone can help, before @MegMll totally loose track of what she had undertaken.

jorisv commented:
Hello @Ipuch,

Now we released pinocchio 4, I will probably be able to work again on this PR next month.
I will also check if there is some other engineer in the team able to work on it.

MegMll committed:
Core - update spline to allow spans to be between min and max, instead of 0 and 1

MegMll committed:
Core - update spline to allow spans to be between min and max, instead of 0 and 1

MegMll commented:
Okay i won't be available in May, but if i need to make some changes myself, i guess it can wait June, anyway

MegMll committed:
core - move joint spline definition

MegMll committed:
core - move joint spline definition

MegMll committed:
core - update splines algorithms

MegMll committed:
core - update splines algorithms

MegMll committed:
core - add spline joint

MegMll committed:
core - add spline joint

MegMll committed:
parser - add spline joint

MegMll committed:
parser - add spline joint

MegMll committed:
core - fix some errors and warnings (some errors due to rebase still remaining)

MegMll committed:
core - fix some errors and warnings (some errors due to rebase still remaining)

MegMll committed:
unittest - use jointModelWithParameters

MegMll committed:
unittest - use jointModelWithParameters

MegMll committed:
unittest - use std::vector

MegMll committed:
unittest - use std::vector

MegMll committed:
remove useless file

MegMll committed:
remove useless file

MegMll committed:
unittest - fix compilation warnings

MegMll committed:
unittest - fix compilation warnings

MegMll committed:
unittest - add spline joint to serialization test

MegMll committed:
unittest - add spline joint to serialization test

Ipuch commented:
Thanks @MegMll for rebasing :+1:

MegMll committed:
unittest - update serialization test to use jointModelWithParameters

MegMll committed:
unittest - update serialization test to use jointModelWithParameters

MegMll committed:
unittest - clean, remove comments

MegMll committed:
unittest - clean, remove comments

MegMll commented:
### To do
- Finish the min/max knot vectors (still need to update constructor to add min/max limits)
- Add multiple knot vector computation (open, closed ...)
- Fit spline algorithm
- Fix ubuntu CI
-  Redo changelog
- precommit

Ipuch committed:
fix cmake: @lucasJoseph

Ipuch commented:
@MegMll and @LucasJoseph pinocchio/algorithm/splines.hpp also need to be added into the sources.cmake

Ipuch committed:
fix: splines.hpp to sources.cmake

MegMll committed:
core : factorize code calc functions

MegMll committed:
core : finish adding min/max for the knot vector + fix knot vector computation (@lpuch)

MegMll committed:
unittest : modify knot unittest to test min/max

MegMll committed:
core : add min/max for spline joint

Ipuch committed:
refactor: decoupling the knotvector scaling

fix: remove extra res.min_q, max_q

Ipuch committed:
tests: joint-generic bounding q with joint limits

jorisv commented on include/pinocchio/src/multibody/joint/spline-utils.hxx:
This file should not be in algorithm: it's not an algorithm we want to directly distribute to the community.
We can put it in `include/pinocchio/src/multibody/joint/splines_utils.hxx` and put struct and functions in the internal namespace (pinocchio::internal).

The casadi specialization can be put in `include/pinocchio/src/autodiff/casadi/multibody/joint/splines_utils.hxx`

jorisv committed:
spline: Apply new header convention

jorisv commented on examples/spline-joint.py:
```diff
@@ -0,0 +1,94 @@
+import time
+
+import hppfcl
```
Use coal instead

jorisv committed:
spline: Fix build

MegMll committed:
core - add a new struct JointSplineBuilder

MegMll committed:
unittest - update tests to use new builder struct

MegMll committed:
model-graph - update jointSpline

MegMll committed:
bindings - update jointSpline bindings

jorisv commented on include/pinocchio/src/multibody/joint/joint-spline.hxx:
```diff
@@ -0,0 +1,697 @@
+      max_q = knotVector[knots.size() - 1];
+      knots = knotVector;
+
+      ctrlFrames = controlFrames;
```
You can copy it in the initialization list

```
    JointModelSplineTpl(
      const std::vector<Transformation_t> & controlFrames,
      const Vector & knotVector,
      const size_t degree)
    : degree(degree)
    , knots(knotVector)
    , ctrlFrames(controlFrames)
```

jorisv commented on include/pinocchio/src/multibody/joint/joint-spline.hxx:
```diff
@@ -0,0 +1,697 @@
+    {
+      assert(
+        check_expression_if_real<Scalar>(qs[0] >= knots[0] && qs[0] <= knots(knots.size() - 1))
+        && "Spline joint configuration (q) must be between 0 and 1. ");
```
```suggestion
      assert(
        check_expression_if_real<Scalar>(qs[0] >= min_q && qs[0] <= max_q)
        && "Spline joint configuration (q) must be between min_q and max_q.");
```

jorisv commented on include/pinocchio/src/multibody/joint/joint-spline.hxx:
```diff
@@ -0,0 +1,697 @@
+    {
+      assert(
+        check_expression_if_real<Scalar>(qs[0] >= knots[0] && qs[0] <= knots(knots.size() - 1))
+        && "Spline joint configuration (q) must be between min and max. ");
```
```suggestion
      assert(
        check_expression_if_real<Scalar>(qs[0] >= min_q && qs[0] <= max_q)
        && "Spline joint configuration (q) must be between min_q and max_q.");
```

jorisv commented on include/pinocchio/src/multibody/joint/joint-spline.hxx:
```diff
@@ -0,0 +1,697 @@
+      return res;
+    }
+
+    void computeRelativeMotions()
```
Should be private

jorisv commented on include/pinocchio/src/multibody/joint/joint-spline.hxx:
```diff
@@ -0,0 +1,697 @@
+        relativeMotions.push_back(log6(ctrlFrames[i].inverse() * ctrlFrames[i + 1]));
+    }
+
+    void computeBasisFunctions(
```
Should be private

jorisv commented on include/pinocchio/src/multibody/joint/joint-spline.hxx:
```diff
@@ -0,0 +1,697 @@
+      }
+    }
+
+    void computeTransformations(
```
Should be private

jorisv commented on include/pinocchio/src/multibody/joint/joint-spline.hxx:
```diff
@@ -0,0 +1,697 @@
+      }
+    }
+
+    Scalar bsplineBasis(size_t i, size_t k, const Scalar x) const
```
`bsplineBasis`, `bsplineBasisDerivative` and `bsplineBasisDerivative2` should be standalone function in `internal` namespace in `spline-utils.hxx`. So they are easily testable and not in public API.

jorisv commented on include/pinocchio/src/multibody/joint/joint-spline.hxx:
```diff
@@ -0,0 +1,697 @@
+      return *this;
+    }
+
+    JointModelSplineBuilderTpl & withRange(const Scalar min_q, const Scalar max_q)
```
Range should be a parameter of `withOpenUniformKnots` and `withUniformKnots`. It's only make sens when we call these functions.

jorisv commented on src/parsers/graph/model-graph-algo.cpp:
```diff
@@ -243,6 +246,14 @@ namespace pinocchio
+            .withControlFrameVector(joint.ctrlFrames)
+            .withDegree(joint.degree)
+            .withRange(joint.min_q, joint.max_q)
+            .build();
```
See `graph::JointSpline` discussion

jorisv commented on include/pinocchio/src/parsers/graph/joints.hxx:
```diff
@@ -248,6 +248,49 @@ namespace pinocchio
       }
     };
 
+    struct JointSpline
```
Hmm, this one is a little hard with the builder…
We can create a `graph::JointSplineBuilder` with the same API than JointModelSplineBuilderTpl.
`graph::JointSplineBuilder` should store all parameters to directly build a `JointModelSplineTpl` with parametrized constructor.
In this case, we should factorize the knot vector computation. You can set the knot vector computation algorithm in `spline-utils.hxx"

jorisv commented on examples/spline-joint.py:
This should be added to `examples/CMakeLists.txt`

jorisv committed:
graph: Explain why we can't reverse the joint

MegMll committed:
core - move splines algos into internal and/or private

jorisv committed:
unittest: Fix joint-init.hpp build

jorisv committed:
unittest: Use joint-init.hpp in visitor.cpp

jorisv committed:
unittest: Use joint-init.hpp in joint-motion-subspace.cpp

jorisv committed:
unittest: Use joint-init.hpp in all-joints.cpp

jorisv committed:
unittest: Remove dead code

jorisv committed:
mpfr: Remove useless include

jorisv committed:
changelog: Set in Unreleased and add example link

jorisv committed:
casadi: Fix scalar casting

jorisv committed:
casadi: Fix expression evaluation

MegMll committed:
core - make test casadi compatible

MegMll committed:
unittest - Fix degree

MegMll committed:
unittest - Make tset compatible with new internal api

jorisv committed:
graph: Add JointSplineBuilder

jorisv committed:
joint: Remove moved methods

jorisv committed:
joint: Remove one warning

jorisv committed:
casadi: Fix build

jorisv committed:
python: Fix binding build and add graph::JointSplineBuilder

jorisv committed:
examples: Use coal and run when running ctest

jorisv committed:
python: Add JointModelSplineBuilder binding

jorisv committed:
lint: Apply clang-format

jorisv committed:
joint: Remove warning

jorisv committed:
unittest: Use builder to build joint

jorisv committed:
unittest: Fix cppad-joints build

MegMll committed:
unittest - rewrite most tests to use internal functions directly

MegMll committed:
core - write correct test for range inside basis function computation

Ipuch committed:
example: knee with spline joint.

Ipuch committed:
example: spline knee joint with better curves.

Ipuch committed:
example: spline knee joint make it minimal

Ipuch committed:
example: parabola in spline-joint.py

example: spline-joint.py x aba with a parabola with @LucasJoseph

jorisv committed:
spline: Fix FindSpan and improve tests

jorisv committed:
test: Relax precision on finite-differences test

jorisv committed:
test: fix some assert and remove warnings

jorisv committed:
spline: Remove more warnings

jorisv committed:
unittest: Use an open uniform spline by default to help numerical precision

jorisv committed:
unittest: Add a bound test to bsplineBasis

jorisv committed:
python: Expose JointDataSpline specialized members

jorisv committed:
warning: Remove a bunch of clang warning

Ipuch commented:
**Current todo**

- [ ] UT findspan (Joris)
- [ ] UT test q negatif (Joris)
- [ ] UT crash outside [0,1], manage that case (Joris)
- [x] What to do at "q_max+1" from theoretical point of view, open an issue (@MegMll)
- [x] Put the video into the header of the PR (@MegMll)
- [x] Give what are the constraints of spline (nb of knotvector points vs degree/order of spline) (@MegMll)
- [x] Give what is the maximum repetition in the knotvector, (verify it's degree +1) (@MegMll)
- [x] Verify the threshold of finitie differences, it's too low for now (@MegMll)

MegMll committed:
core - add Continuity check over knot vector

MegMll committed:
unittest - add test for basis function continuity

MegMll committed:
unittest - fix unittest for basis function derivatives (finite difference errors were amplified by small differences between knot vectors values)

MegMll committed:
examples - apply linting

jorisv committed:
spline: Add deBoor algorithm to evaluate the BSpline

jorisv committed:
spline: Remove warnings

jorisv committed:
spline: Implement deBoorCumBasisSparse

jorisv committed:
spline: Rename deBoorBasis into deBoor

jorisv committed:
spline: Add bsplineBasis to compute all basis function with minimal computation

jorisv committed:
spline: Remove other deBoor like algorithm

jorisv committed:
spline: Add comments

jorisv committed:
spline: First working version with deBoorBasis

jorisv committed:
spline: Add standalone function to compute cumulative bspline derivatives

jorisv committed:
spline: Move bsplineBasis in unit tests and test cumulativeBasisDerivative

jorisv committed:
spline: Fix build

jorisv committed:
spline: Add deBoorFullBasis function

jorisv committed:
spline: Improve doc and comments

jorisv committed:
spline: Add Full variant to support casadi

jorisv committed:
spline: Refactor code to implement spline transformation on minimal and full basis vector

jorisv committed:
spline: Add computeSplineKinematicsFull

jorisv committed:
spline: Fix changelog

jorisv committed:
spline: Test computeSplineKinematicsFull.

Improve derivative test by running them on order 3 spline.

jorisv committed:
Spline: Setup SplineKinematics structure
This allow to use algorithms dedicated to some scalar types.

jorisv committed:
spline: Try to fix Windows build issue

jorisv committed:
examples: Add guards to stop example if meshcat is not installed

jorisv committed:
spline: Remove warnings

jorisv committed:
spline: Improve assert

jorisv committed:
spline: Rename deBoorFulBasis into deBoorBasisFull

jorisv committed:
spline: Store deBoorBasis in a dense vector

jorisv committed:
spline: Enforce to evalute spline in valid range

jorisv committed:
python: Add attributes

jorisv committed:
spline: Remove useless macro

jorisv committed:
spline: Fix check_expression_if_real use

jorisv committed:
unittest: Use degree 3 spline

jorisv committed:
changelog: List second example

jorisv committed:
unittest: Use joint-init.hpp in cppad/joints.cpp

jorisv committed:
spline: Add spline constructor guards test
